Add new AuxKernels and AuxScalarKernels for calculation of Henry and mass transfer coefficient values

bobsalko opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
The mass transfer modeling work that supports modeling of molten salt reactors and fission gas transfer requires calculation of Henry coefficients and mass transfer coefficients.  Models have been developed for that calculations as part of MSR modeling efforts in NEAMS.  This MR will merge these new capabilities into MOOSE.

## Design
<!--A concise description (design) of the enhancement.-->

A set of AuxKernels and AuxScalarKernels have been added for the Henry coefficient calculation as well as the mass transfer coefficient calculation.  The Henry coefficient kernels include models for both FLIBE and FLINAK salts as well as an option for the user to enter a custom model.  The mass transfer coefficient model includes options using the Wilke-Chang and Stokes-Einstein models.  Tests have been added for all kernels to ensure the values are being correctly calculated.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
This will add new AuxKernels and AuxScalarKernels that other MOOSE applications can utilize for modeling of mass transfer.


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

bobsalko committed:
Add Henry model to MOOSE
refs #32104

bobsalko committed:
Add MTC AuxKernels

bobsalko committed:
Autoformat

bobsalko commented:
added issue to track: #32106

GiudGiud commented:
@bobsalko who are you working with on this? To see if there is a preferred reviewer

wgurecky commented:
> @bobsalko who are you working with on this? To see if there is a preferred reviewer

This MR is to satisfy a milestone for a WP in SM&C. @tanoret

tanoret commented:
Nice contribution!

I would say that the transfer kernel `GasLiquidMassTransferAux` actually belongs to scalar_transport or navier_stokes, as @wgurecky previously mentioned. The mass transfer by partial pressures is a physical process, without chemical reactions, i.e. electrons transferred involved. So, chemical_reactions could be a bit confusing.

As for `HenryGasConstantAux`, I agree with you guys that `FluidProperties` may be a better choice.

What do you think?

GiudGiud commented on modules/chemical_reactions/include/auxkernels/GasLiquidMassTransferAux.h:
```diff
@@ -0,0 +1,54 @@
+#pragma once
```
missing copyright

GiudGiud commented on modules/chemical_reactions/src/auxkernels/GasLiquidMassTransferAux.C:
```diff
@@ -0,0 +1,99 @@
+#include "GasLiquidMassTransferAux.h"
```
same comment for copyright in source

GiudGiud commented on modules/chemical_reactions/src/auxkernels/GasLiquidMassTransferAux.C:
```diff
@@ -0,0 +1,99 @@
+#include "MathUtils.h"
+#include "SinglePhaseFluidProperties.h"
+
+registerMooseObject("MooseApp", GasLiquidMassTransferAux);
```
register this to the ScalarTransportApp if in scalar_transport module, FLuidPropertiesApp if in Fluid_properties etc
otherwise doco check will fail

GiudGiud commented on modules/chemical_reactions/src/auxkernels/GasLiquidMassTransferAux.C:
```diff
@@ -0,0 +1,99 @@
+  // Required params
+  params.addRequiredCoupledVar("p", "Pressure [Pa]");
+  params.addRequiredCoupledVar("T", "Temperature [K]");
+  params.addRequiredCoupledVar("fluid_velocity", "Velocity vector [m/s]");
```
Says vector here but it is being stored in a single-component variable value ?

GiudGiud commented on modules/chemical_reactions/src/auxkernels/GasLiquidMassTransferAux.C:
```diff
@@ -0,0 +1,99 @@
+      // 1955, 1(2) 264-270 Units of Wilke Chang are g-cm-s
+      Real kg_to_g = 1000.0;
+      Real m_to_cm = 100.0;
+      Real cm_to_m = 1 / m_to_cm;
```
```suggestion
      Real cm_to_m = 1. / m_to_cm;
```

GiudGiud commented on modules/chemical_reactions/src/auxkernels/GasLiquidMassTransferAux.C:
```diff
@@ -0,0 +1,99 @@
+      Real molar_weight = _mw * kg_to_g;               // g/mol
+      Diffusivity = _wc * _temperature[_qp] * std::sqrt(_phi * molar_weight) /
+                    (mu_cgs * std::pow(molar_volume, 0.6));
+      Diffusivity = Diffusivity * pow(cm_to_m, 2);
```
```suggestion
      Diffusivity = Diffusivity * Utility::pow<2>(cm_to_m);
```

GiudGiud commented on modules/chemical_reactions/src/auxkernels/HenryGasConstantAux.C:
```diff
@@ -0,0 +1,88 @@
+  Real value =
+      _alpha * 4 * libMesh::pi * std::pow(_radius * m_to_ang, 2) *
+          (_gamma_0 + _dgamma_dT * (_temperature[_qp] - 273.15)) +
+      _beta * 4 / 3 * libMesh::pi * std::pow(_radius * m_to_ang, 3) * _Rgas * _temperature[_qp];
```
```suggestion
      _beta * 4. / 3. * libMesh::pi * std::pow(_radius * m_to_ang, 3) * _Rgas * _temperature[_qp];
```
careful with those. 4 / 3 in C++ is integer division

GiudGiud commented on modules/chemical_reactions/src/auxscalarkernels/HenryGasConstantScalarAux.C:
```diff
@@ -0,0 +1,88 @@
+    case Saltlist::CUSTOM:
+      // Ensure that all parameters were set by user
+      if (!isParamSetByUser("alpha"))
+        mooseError("Must include alpha parameter when using custom salt Henry gas model");
```
```suggestion
        paramError("alpha", "Must include alpha parameter when using custom salt Henry gas model");
```

GiudGiud reviewed (changes requested):
which code is going to use this? An auxkernel does not allow the propagation of derivative terms, so it prevents getting a perfect Jacobian when the problem is solved in a fully coupled approach. Depending on the code you want to use this with, I might recommend material properties, functor properties, or auxkernels if (linearFV, no Jacobian there)

If the equations are coupled only 1 way (p,T,v influence transport, but transport does not influence them), we might be looking at a segregated multi-system solve in which case auxkernels would be fine too

bobsalko commented:
> which code is going to use this? An auxkernel does not allow the propagation of derivative terms, so it prevents getting a perfect Jacobian when the problem is solved in a fully coupled approach. Depending on the code you want to use this with, I might recommend material properties, functor properties, or auxkernels if (linearFV, no Jacobian there)

That is a good point about the derivative term.  If we put it into the EOS class (i.e., `SalineMoltenSaltFluidProperties`, and maybe also `FlibeFluidProperties` and `FlinakFluidProperties`), then the derivative terms can be retrieved.  I'm not sure about your comment about making it material properties or functor properties.  What would be the benefit there over EOS functions?

Also, I think the Henry coefficient makes sense for EOS potentially, but what about the mass transfer coefficient calculation?  Would that be more suitable for `AuxKernels`?

GiudGiud commented:
> I'm not sure about your comment about making it material properties or functor properties. What would be the benefit there over EOS functions?

The problem with EOS (FluidProperties) is that the routines are not very general. The calculation of the coefficients is only for a few fluids (not all the single phase ones) and is one model among others, so the parameter list would need to grow every time we add a new model.
I could see it done within the EOS, but with a slightly different design. Create a helper class that adds the parameters and the coefficient routines, and add that helper to the 3 fluids of interest. Then when using it, you will first cast the fluid properties to the helper class. 

On the other hand, a material is better self-contained. It's very close to the auxkernel design, it's just more efficient because it's computed on the fly rather than storing a whole field for each coefficient.

wgurecky commented:
@GiudGiud  and @tanoret Thanks for the feedback and discussion!  I'm in agreement that migrating to the Henry's coefficient callculation for high volatile, noble gases in molten salts over to the MOOSE Material system is a good strategy.  The Henry's coefficient is indeed depedant on temperature and, weakly, on pressure (through mu and rho).

Should the present auxkernels/GasLiquidMassTransferAux   likewise be migrated over to the Material system as well?  The mass transfer coefficients are a function of the local fluid velocity, and temperature, so I could also see how accounting jacobian contributions from these could help in some cases.

bobsalko committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

bobsalko committed:
Correct header info

bobsalko committed:
Change Henry coefficient to UserObject and add tests

bobsalko committed:
Switched to UserObject for mass transfer coefficient

bobsalko committed:
Switch to relative tolerance

bobsalko committed:
autoformat

GiudGiud commented on modules/scalar_transport/unit/src/GasLiquidMassTransferTest.C:
```diff
@@ -0,0 +1,51 @@
+
+const double tol = 1.0e-6;
+
+const double p = 1e5;
```
```suggestion
const Real tol = libMesh::TOLERANCE;

const Real p = 1e5;
```
etc

that way all the tests scale with the floating point precision MOOSE is compiled with

bobsalko committed:
Address review:
- single line functions in header
- docstring formatting
- remove makefile changes due to dependency between modules
- unit test custom coefficient henry constant
- tighten tolerances

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

bobsalko commented on modules/modules.mk:
```diff
@@ -126,10 +126,15 @@ ifeq ($(XFEM),yes)
 
+ifeq ($(CHEMICAL_REACTIONS),yes)
+        FLUID_PROPERTIES            := yes
+endif
```
Done

bobsalko reviewed (commented):
Most comments addressed.  One question on my end.

bobsalko committed:
Expand unit test

GiudGiud commented:
my contributions were minor (commit history, formatting / precheck)

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3666431/), step Results summary on 294bb64 wanted to post the following:

## Framework test summary

Compared against 24dc6c8 in job [civet.inl.gov/job/3662184](https://civet.inl.gov/job/3662184).

No change
## Modules test summary

Compared against 24dc6c8 in job [civet.inl.gov/job/3662184](https://civet.inl.gov/job/3662184).

No change

GiudGiud committed:
Unrelated unit test fix

bobsalko committed:
Add docs for new user objects

GiudGiud committed:
Documentation fixups

GiudGiud commented:
Thanks for the contribution!
